keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,179 @@
+
+In addition, due to the self-healing nature of Kubernetes, regressions wherein latencies are affected but the overall task is eventually accomplished are not uncommon. With our current monitoring architecture, these "soft regressions" are often difficult to observe and understand. Collecting structured trace data on per-object latencies would enable us to detect these long-term regressions automatically, and quickly determine their root causes.
+
+
+### Goals
+
+* Make it possible to visualize the progress of objects across distinct Kubernetes components
+* Streamline the tedious processes involved in debugging latency issues in Kubernetes
+* Make it possible to identify high-level latency regressions, and attribute them back to offending processes
+
+
+### Non-Goals
+
+* To replace existing logging, metrics, or the events API
+* To trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+
+## Proposal
+
+### Architecture
+
+#### Trace lifecycle
+
+Kubernetes is unique in that it is constantly reconciling its actual state towards some desired state. As a result, it has no definitive concept of an "operation", which breaks the traditional model for distributed tracing. This raises the question of when to begin traces, and when to end them.
+
+In this proposal, we choose to _only_ trace phases of an object's lifecycle wherein it's correcting from an undesired state to its desired state, and to end the trace when it enters this desired state. This means that the same object will export traces for each reconciliation it undergoes. This decision was made because:
+
+1) These reconciliation phases are where Kubernetes performs actual work that can be traced
+2) Prevents traces from spanning the entire life of an object, which can be indefninite
+3) Provides a consistent, concrete duration for the root span which remains applicable across object types
+4) The time required for Kubernetes controllers to reconcile desired and actual state is, by definition, a measurement of either latency experienced by end-users in the fulfillment of their requests, or down-time due to a disruption
+
+#### Root spans
+
+In the standard model for distributed tracing, there exists a span in each trace that all other spans are descendents of and which extends the length of the entire trace, called the `root span`. 
+
+The Kubernetes component that kicks off an operation might not be the same component that ends it. In this proposal, when we are at the point where we want to end a root span, we craft a span to export which acts as the root span for the trace. For example, when the kubelet updates a pod from `Pending` to `Running`, it creates a root span using the start time of the pod as the start, and the current time as the end.
```

Comment:
What about cases when the component that is finishing doesn't know the time when it was started?
As an example, the action may be triggered by updating an object, and we generally don't persistent anywhere information about when the object was update (even last one).

Reply:
Root spans are useful to have, but not critical for using tracing.  Essentially what you get by adding a root span is being able to collapse the entire trace during visualizations, as all spans have a common parent.  Tracing backends still calculate the total duration of all spans.

The current plan for alpha is to add root spans where possible (creation, deletion), and not where it isn't (update, reconcile).